casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala

```diff
@@ -0,0 +1,131 @@
+    val start: Long,
+    val duration: Long,
+    private val sigsF: Ref[F, Set[Signature]]) {
+  private implicit val logSource: LogSource = LogSource(this.getClass)
+
+  val candidate                         = ApprovedBlockCandidate(Some(block), requiredSigs)
+  private val u                         = UnapprovedBlock(Some(candidate), start, duration)
+  private val serializedUnapprovedBlock = u.toByteString
+  private val candidateHash             = PrettyPrinter.buildString(block.blockHash)
+  private val sigData                   = Blake2b256.hash(candidate.toByteArray)
+
+  def addApproval(a: BlockApproval): F[Unit] = {
+    val validSig = for {
+      c   <- a.candidate if c == this.candidate
+      sig <- a.sig if Validate.signature(sigData, sig)
+    } yield sig
+    val sender = a.sig.fold("<Empty Signature>")(sig => Base16.encode(sig.publicKey.toByteArray))
+
+    validSig.fold(Log[F].warn(s"APPROVAL: ignoring invalid block approval from $sender")) { sig =>
+      sigsF.update(_ + sig) *> Log[F].info(s"APPROVAL: received block approval from $sender")
+    }
+  }
+
+  def currentSigs: F[Set[Signature]] = sigsF.get
+
+  def approvedBlock: F[Option[ApprovedBlock]] =
+    sigsF.get.map(
+      sigs =>
+        if (sigs.size >= requiredSigs) Some(ApprovedBlock(Some(candidate), sigs.toSeq))
+        else none[ApprovedBlock])
+
+  //TODO: potential optimization, only send to peers we have not
+  //      reveived a valid signature from yet
+  def sendUnapprovedBlock: F[Unit] =
+    for {
+      _ <- Log[F].info(s"APPROVAL: Beginning send of UnapprovedBlock $candidateHash to peers...")
+      _ <- CommUtil.sendToPeers[F](transport.UnapprovedBlock, serializedUnapprovedBlock)
+      _ <- Log[F].info(s"APPROVAL: Sent UnapprovedBlock $candidateHash to peers.")
+    } yield ()
+
+  def sendApprovedBlock: F[Boolean] =
+    approvedBlock.flatMap {
+      case Some(a) =>
+        val serializedApprovedBlock = a.toByteString
+        for {
+          _ <- Log[F].info(s"APPROVAL: Beginning send of ApprovedBlock $candidateHash to peers...")
+          _ <- CommUtil.sendToPeers[F](transport.ApprovedBlock, serializedApprovedBlock)
+          _ <- Log[F].info(s"APPROVAL: Sent ApprovedBlock $candidateHash to peers.")
+        } yield true
+
+      case None => false.pure[F]
+    }
+
+}
+
+object ApproveBlockProtocol {
+
+  def apply[F[_]](implicit instance: ApproveBlockProtocol[F]): ApproveBlockProtocol[F] = instance
+
+  def create[F[_]: Capture: Sync: NodeDiscovery: TransportLayer: Log: Time: ErrorHandler](
+      block: BlockMessage,
+      requiredSigs: Int,
+      duration: Long): F[ApproveBlockProtocol[F]] =
+    for {
+      now   <- Time[F].currentMillis
+      sigsF <- Ref.of[F, Set[Signature]](Set.empty[Signature])
+    } yield new ApproveBlockProtocol(block, requiredSigs, now, duration, sigsF)
+
+  def blockApprovalPacketHandler[F[_]: Monad: ApproveBlockProtocol](
+      peer: PeerNode): PartialFunction[Packet, F[Option[Packet]]] =
```

Comment:
`peer` is not used. It seems like an oversight?

Reply:
Nope. Not all `PacketHandler`s need to know the peer that the packet came from, but some do, e.g. https://github.com/rchain/rchain/blob/dev/casper/src/main/scala/coop/rchain/casper/util/comm/BlockApproverProtocol.scala#L50
For this reason the effect constructor includes the peer https://github.com/rchain/rchain/blob/dev/comm/src/main/scala/coop/rchain/p2p/effects/PacketHandler.scala#L25

Reply:
But you don't use `peer` here and don't make a use of https://github.com/rchain/rchain/blob/dev/comm/src/main/scala/coop/rchain/p2p/effects/PacketHandler.scala#L25

Reply:
Well I hadn't yet, but was planning on it. If you will take a different approach then you can drop the `peer` parameter